syntax = "proto2";

package devtools_crosstool_autofdo;


// Next Available: 12.
message PropellerOptions {
  // binary file name.
  optional string binary_name = 1;

  // perf.data file name.
  repeated string perf_names = 2;

  // propeller cluster information output file name.
  optional string cluster_out_name = 3;

  // propeller symbol order output file name.
  optional string symbol_order_out_name = 4;

  // profiled binary name; if not empty, use this as file name to match MMAP
  // events in perf.data file, can be "".
  optional string profiled_binary_name = 5;

  // ignore the build id; if set, use binary_name or profiled_binary_name to
  // match MMAP events. Otherwise, use buildid (if exists) to match.
  optional bool ignore_build_id = 6;

  // Internal options used by unittest to keep frontend data.
  optional bool keep_frontend_intermediate_data = 7;

  // Parameters used to calculate the Extended TSP score for layout.
  optional PropellerCodeLayoutParameters code_layout_params = 8;

  // Include extra information such as per-function layout scores in the
  // propeller cluster file.
  optional bool verbose_cluster_output = 9 [default = false];

  // Directory path for dumping the (hot) cfgs in dot format. Dumping won't be
  // done if field is unset.
  optional string cfg_dump_dir_name = 10;

  // Start a http-server to handle /statusz.
  optional bool http = 11 [default = false];
}

// Next Available: 13.
message PropellerCodeLayoutParameters {
  optional uint32 fallthrough_weight = 1 [default = 10];
  optional uint32 forward_jump_weight = 2 [default = 1];
  optional uint32 backward_jump_weight = 3 [default = 1];
  optional uint32 forward_jump_distance = 4 [default = 1024];
  optional uint32 backward_jump_distance = 5 [default = 680];
  // Whether splitting will be done when merging chains.
  optional bool chain_split = 6 [default = false];
  // Maximum chain size for which splitting is considered at every node bundle.
  optional uint32 chain_split_threshold = 7 [default = 0];
  // Whether to use the call chain clustering algorithm to reorder functions.
  optional bool call_chain_clustering = 8 [default = false];
  // Maximum cluster size for call chain clustering (No further clustering is
  // done on a cluster when it exceeds this size).
  optional uint32 cluster_merge_size_threshold = 9 [default = 2097152];
  // Whether to split the cold blocks in hot functions. When `false` cold blocks
  // are placed immediately after hot ones.
  optional bool split_functions = 10 [default = true];
  // Whether to change the relative order of hot blocks.
  optional bool reorder_hot_blocks = 11 [default = true];
  // Whether to do inter-procedural reordering.
  optional bool inter_function_reordering = 12 [default = false];
}
